posts/2020-02-14-cats_stm_0.7.0.md

```diff
@@ -0,0 +1,110 @@
+or at least some of the docs for this library :D However, I will endeavour to
+give a very brief overview here.
+
+The motivating observation is that the traditional tools for doing concurrent
+programming (mutexes, semaphores, etc) do not compose. Given two functions `f: A
+=> B` and `g: B => C` each acquiring mutexes, we cannot reason about the locking
+behaviour of `f andThen g` in the presence of concurrency (see sections 2.1 and 2.2
+of the paper).
+
+The solution to this is the `STM` monad. This is compositional by definition
+(indeed, this practically _is_ the definition of a monad) and exposes combinators
+such as
+
+`def map[A, B](s: STM[A], f: A => B): STM[B]`
+`def orElse[A](attempt: STM[A], fallback: STM[A]): STM[A]`
+`def check(check: => Boolean): STM[Unit]`
+
+How do we obtain a value of type `STM[A]` in the first place? These are returned by
+the operations defined on `TVar`s (transactional vars):
+`get: STM[A]`
+`set(a: A): STM[Unit]`
+`modify(f: A => A): STM[Unit]`
+
+A value of type `STM[A]` represents a computation that we would like to run
+atomically and which should return a value of type `A`. How do we do that?
+The clue's in the name!
+
+`atomically[F[_], A](s: STM[A]): F[A]`
+
+Why does this return something of type `F[A]` (assume this is `IO[A]` for
+simplicity's sake) rather than something of type `A`? The execution of
+`atomically` is the point at which we acquire locks and mutate `TVar`s ie
+perform side effects and hence must be suspended in `IO`
+
+Here is a contrived example of what this looks like in practice:
+
+```scala
+import cats.effect.{ExitCode, IO, IOApp}
+import io.github.timwspence.cats.stm.{TVar, STM}
+
+object Main extends IOApp {
+
+  override def run(args: List[String]): IO[ExitCode] = for {
+    accountForTim   <- TVar.of[Long](100).commit[IO]
+    accountForSteve <- TVar.of[Long](0).commit[IO]
+    _               <- printBalances(accountForTim, accountForSteve)
+    _               <- giveTimMoreMoney(accountForTim).start
+    _               <- transfer(accountForTim, accountForSteve)
+    _               <- printBalances(accountForTim, accountForSteve)
+  } yield ExitCode.Success
+
+  private def transfer(accountForTim: TVar[Long], accountForSteve: TVar[Long]): IO[Unit] = for {
+    _ <- STM.atomically[IO] {
+      for {
+        balance <- accountForTim.get
+        _       <- STM.check(balance > 100)
+        _       <- accountForTim.modify(_ - 100)
+        _       <- accountForSteve.modify(_ + 100)
+      } yield ()
+    }
+  } yield ()
+
+  private def giveTimMoreMoney(accountForTim: TVar[Long]): IO[Unit] = for {
+    _ <- IO(Thread.sleep(5000))
+    _ <- STM.atomically[IO](accountForTim.modify(_ + 1))
+  } yield ()
+
+  private def printBalances(accountForTim: TVar[Long], accountForSteve: TVar[Long]): IO[Unit] = for {
+    _ <- accountForTim.get.commit[IO].flatMap(b => IO(println(s"Tim: $b")))
+    _ <- accountForSteve.get.commit[IO].flatMap(b => IO(println(s"Tim: $b")))
```

Comment:
Should be `"Steve: $b"` here, right? I'd be inclined to rewrite this as:
```scala
  for {
    amountForTim <- accountForTim.get.commit[IO]
    _              <- IO(println(s"Tim: $b"))
    amountForSteve <- accountForSteve.get.commit[IO]
    _              <- IO(println(s"Steve: $b"))
  }
```
…to avoid mixing `for` and `flatMap`, but that's a matter of taste.